agentops/llms/llama_stack_client.py

```diff
@@ -0,0 +1,235 @@
+                            else:
+                                accum_tool_delta = delta
+                        elif (chunk.event.payload.step_type == "inference" and chunk.event.payload.tool_call_delta and chunk.event.payload.tool_call_delta.parse_status == "success"):
+                            tool_event.name = "ToolExecution - success"
+                            tool_event.params["completion"] = accum_tool_delta
+                            self._safe_record(session, tool_event)
+                        elif (chunk.event.payload.step_type == "inference" and chunk.event.payload.tool_call_delta and chunk.event.payload.tool_call_delta.parse_status == "failure"):
+                            self._safe_record(session, ErrorEvent(trigger_event=tool_event, exception=Exception("ToolExecution - failure")))
+
+                    elif chunk.event.payload.event_type == "step_complete":
+                        if (chunk.event.payload.step_type == "inference"):
+                            llm_event.prompt = [
+                                {"content": message['content'], "role": message['role']} for message in kwargs["messages"]
+                            ]
+                            llm_event.agent_id = check_call_stack_for_agent_id()
+                            llm_event.model = metadata.get("model_id", "Unable to identify model")
+                            llm_event.prompt_tokens = None
+                            llm_event.completion = accum_delta or kwargs["completion"]
+                            llm_event.completion_tokens = None
+                            llm_event.end_timestamp = get_ISO_time()
+                            self._safe_record(session, llm_event)
+                        elif (chunk.event.payload.step_type == "tool_execution"):
+                            tool_event.name = "ToolExecution - complete"
+                            tool_event.params["completion"] = accum_tool_delta
+                            self._safe_record(session, tool_event)
+                    elif chunk.event.payload.event_type == "turn_complete":
+                        pass
+
+                except Exception as e:
+                    self._safe_record(session, ErrorEvent(trigger_event=llm_event, exception=e))
+
+                    kwargs_str = pprint.pformat(kwargs)
+                    chunk = pprint.pformat(chunk)
+                    logger.warning(
+                        f"Unable to parse a chunk for LLM call. Skipping upload to AgentOps\n"
+                        f"chunk:\n {chunk}\n"
+                        f"kwargs:\n {kwargs_str}\n"
+                    )
+
+            if kwargs.get("stream", False):
+                def generator():
+                    for chunk in response:
+                        handle_stream_chunk(chunk)
+                        yield chunk
+                return generator()
+            elif inspect.isasyncgen(response):
+                async def async_generator():
```

Comment:
this may cause race conditions because:

- it's an sync context
- computes values on the fly by aggregating on top of the previous values, which in a concurrent state can lead to inconsistencies. Might want to use `asyncio.Lock`

Reply:
but only testing tells, im 80% confident about the possibility, just by looking at it

Reply:
I don't understand this

Reply:
https://stackoverflow.com/questions/25799576/whats-python-asyncio-lock-for